python/lsst/ctrl/bps/bps_core.py

```diff
@@ -388,26 +390,88 @@ def _update_task(self, task_abbrev, tnode, qlfn):
+        Assumes initialization jobs must be new (and only) source in workflow graph.
         """
-        task_abbrev_list = [x.strip() for x in self.pipeline]
-        for abbrev_id, task_abbrev in enumerate(task_abbrev_list, 0):
-            if abbrev_id != 0:
-                # get current task's init task node
-                st_node_name = init_nodes[task_abbrev][TASKNODE]
-
-                # get previous task's init output file node
-                prev_abbrev = task_abbrev_list[abbrev_id - 1]
-                sf_node_name = init_nodes[prev_abbrev][FILENODE]
+        # Create a workflow graph that will have task and file nodes necessary for
+        # initializing the pipeline execution
+        init_graph = self._create_workflow_init_graph()
+        _LOG.debug("init_graph nodes = %s", init_graph.nodes())
+
+        # Find source nodes in workflow graph.
+        task_nodes = [n for n, d in self.gen_wf_graph.nodes(data=True) if d["node_type"] == TASKNODE]
+        task_graph = bipartite.projected_graph(self.gen_wf_graph, task_nodes)
+        task_sources = [n for n in task_graph if task_graph.in_degree(n) == 0]
+        _LOG.debug("workflow sources = %s", task_sources)
+
+        # Find sink nodes of initonly graph.
+        init_sinks = [n for n in init_graph if init_graph.out_degree(n) == 0]
+        _LOG.debug("init sinks = %s", init_sinks)
+
+        # Add initonly nodes to Workflow graph and make new edges.
+        self.gen_wf_graph.add_nodes_from(init_graph.nodes(data=True))
+        self.gen_wf_graph.add_edges_from(init_graph.edges())
+        for source in task_sources:
+            for sink in init_sinks:
+                self.gen_wf_graph.add_edge(sink, source)
+
+    def _create_workflow_init_graph(self):
+        """Create workflow subgraph for running initialization job(s).
+        """
+        _LOG.info("creating init subgraph")
+        initgraph = networkx.DiGraph()
+
+        # create nodes for executing --init-only
+        tnode_name = f"pipetask_init"
+        lfn = "pipetask_init"
+        initgraph.add_node(
+            tnode_name,
+            node_type=TASKNODE,
+            task_abbrev="pipetask_init",
+            label="pipetask_init",
+            job_attrib={"bps_jobabbrev": "pipetask_init"},
+            shape="box",
+            fillcolor="gray",
+            # style='"filled,bold"',
+            style="filled",
+        )
+        self._update_task("pipetask_init", initgraph.nodes[tnode_name], self.qgraph_filename)
+
+        _LOG.info("creating init task input(s)")
+        lfn = basename(self.qgraph_filename)
+        qnode_name = lfn
+        initgraph.add_node(
+            qnode_name,
+            node_type=FILENODE,
+            lfn=lfn,
+            label=lfn,
+            pfn=self.qgraph_filename,
+            ignore=False,
+            data_type="quantum",
+            shape="box",
+            style="rounded",
+        )
+        initgraph.add_edge(qnode_name, tnode_name)
```

Comment:
The code for init was kept similar to the main code.  qnode_name was purposefully picked because it stood for QuantumGraph pickle file.  But changing it to fnode_name.